server/athenian/api/internal/miners/github/pull_request.py

```diff
@@ -3103,3 +3106,28 @@ def _validate(self, facts: PullRequestFacts, url: str) -> None:
                 facts.released - facts.merged,
             )
             raise ImpossiblePullRequest()
+
+
+async def prs_node_id_to_number(
+    node_ids: Sequence[int],
+    meta_ids: tuple[int, ...],
+    mdb: DatabaseLike,
+) -> dict[int, int]:
+    """Build a mapping to convert PR node identifier to PR number.
+
+    The mapping will not include not found node identifiers.
+    """
+    selects = []
+    node_id_cond, hints = column_values_condition(NodePullRequest.node_id, node_ids)
+    selects = [
+        sa.select(NodePullRequest.node_id, NodePullRequest.number).where(
+            NodePullRequest.acc_id == meta_id, node_id_cond,
+        )
+        for meta_id in meta_ids
+    ]
+    stmt = sa.union_all(*selects)
+
+    for hint in hints:
+        stmt = stmt.with_statement_hint(hint)
+    rows = await mdb.fetch_all(stmt)
+    return dict(rows)
```

Comment:
therefore,
```suggestion
    return dict(df[NodePullRequest.node_id.name].values, df[NodePullRequest.number.name].values)
```

Reply:
do you mean dict(zip(... values, ...values)) ?
I've benchmarked this locally with ~600 rows and
```
async def simple(mdb):
    rows = await mdb.fetch_all(stmt)
    return dict(rows)
```
outperforms
```
async def complex_(mdb):
    df = await read_sql_query(stmt, mdb, columns)
    return dict(
        zip(df[NodePullRequest.node_id.name].values, df[NodePullRequest.number.name].values),
    )
```
7.5 ms simple, 8.5 ms complex.
I've also tried `return dict(df.itertuples(index=False, name=None))`
Is there maybe a faster  way to do the df => dict conversion ?

Reply:
Nah, local benchmarks are all lies. The vCPUs we are using in GCP make the profile much more skewed toward CPU. As a rule of thumb, if the returned columns are types, always try to prefer `read_sql_query` instead of the vanilla asyncpg rows.

Reply:
BTW, I've just realized: you shouldn't make the dict at all! The logic must be vectorized downstream using `np.searchsorted`

Reply:
Another rule of thumb: avoid Python objects like a plague, preferring vectorized numpy arrays whereever possible.